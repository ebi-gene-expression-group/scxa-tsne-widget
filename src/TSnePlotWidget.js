import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'

import TSnePlotView from 'expression-atlas-experiment-page-tsne-plot'

class TSnePlotWidget extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      plotdata: {
        metadata: [],
        perplexities: []
      },
      selectedColourBy: `inferred_cell_type`,
      selectedColourByCategory: `metadata`,
      metadataErrorMessage: null,
      loadingMetadata: false,
    }
  }

  async _fetchAndSetState(resource, baseUrl, dataField, errorMessageField, loadingField) {
    this.setState({
      [loadingField]: true
    })

    const url = URI(resource, baseUrl).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      this.setState({
        [dataField]: await response.json(),
        [errorMessageField]: null,
        [loadingField]: false,
      })

      if (!(this.state.plotdata.metadata.indexOf(`inferred_cell_type`) > -1)) {
        this.setState({
          selectedColourBy: this.state.plotdata.metadata[0].value //selects first element of metadata array if inferred_cell_type is not present
        })
      }

    } catch (e) {
      this.setState({
        [errorMessageField]: `${e.name}: ${e.message}`,
        [loadingField]: false
      })
    }
  }

  _fetchAndSetStateMetadata(
    {atlasUrl, experimentAccession}) {
    const resource = `json/experiments/${experimentAccession}/metadata/tsneplot`

    this._fetchAndSetState(
      resource, atlasUrl, `plotdata`, `metadataErrorMessage`, `loadingMetadata`) // this will work once backend code is merged in sc atlas. In meantime, it will not fetch metadata option as there is no endpoint currently in sc atlas
  }

  _onChangeColourBy(colourByCategory, colourByValue) {
    this.setState({
      selectedColourBy: colourByValue,
      selectedColourByCategory: colourByCategory
    })
  }

  componentDidMount() {
    this._fetchAndSetStateMetadata(this.props)
  }

  render() {
    const {height, atlasUrl, suggesterEndpoint} = this.props
    const {wrapperClassName, clusterPlotClassName, expressionPlotClassName} = this.props
    const {geneId, speciesName, ks, experimentAccession} = this.props
    const {metadata, perplexities} = this.state.plotdata
    const {selectedColourBy, selectedColourByCategory, loadingMetadata} = this.state

    const perplexitiesOrdered = perplexities.sort((a, b) => a - b)

    return (
      loadingMetadata ?
        <p className={`row column loading-message`}>Loading, please waitâ€¦</p> :
        <React.Fragment>
          <TSnePlotView
            atlasUrl={atlasUrl}
            suggesterEndpoint={suggesterEndpoint}
            wrapperClassName={wrapperClassName}
            clusterPlotClassName={clusterPlotClassName}
            expressionPlotClassName={expressionPlotClassName}
            speciesName={speciesName}
            experimentAccession={experimentAccession}
            ks={ks} // for future use
            metadata={metadata}
            selectedColourBy={selectedColourBy} // used default k value as we are not showing widget controls for now
            selectedColourByCategory={selectedColourByCategory} // Is the plot coloured by clusters or metadata (as we not showing controls so used 'clusters' as default selected category
            perplexities={perplexities}
            selectedPerplexity={perplexitiesOrdered[Math.round((perplexitiesOrdered.length - 1) / 2)]} //default value given for now
            geneId={geneId}
            height={height}
            onChangeColourBy={(colourByCategory, colourByValue) =>
              this._onChangeColourBy(colourByCategory, colourByValue)
            }
            showControls={false} // flag to control weather controls over tsne plots are shown or not
          />
          <p>
            To know more about this experiment please go to <a target={`_blank`} href={`https://www.ebi.ac.uk/gxa/sc/experiments/${experimentAccession}`}>Single Cell Expression Atlas </a>.
          </p>
        </React.Fragment>
    )
  }
}

TSnePlotWidget.propTypes = {
  atlasUrl: PropTypes.string,
  wrapperClassName: PropTypes.string,
  clusterPlotClassName: PropTypes.string,
  expressionPlotClassName: PropTypes.string,
  experimentAccession: PropTypes.string.isRequired,
  geneId: PropTypes.string.isRequired,
  ks: PropTypes.arrayOf(PropTypes.number),
  suggesterEndpoint: PropTypes.string,
  speciesName: PropTypes.string,
  height: PropTypes.number
}


TSnePlotWidget.defaultProps = {
  atlasUrl: `http://localhost:8080/gxa/sc/`,      //no options will be shown until new metadata endpoint is implemented in prod/ to test change it to localhost
  suggesterEndpoint: `json/suggestions`,
  wrapperClassName: `row expanded`,
  clusterPlotClassName: `small-12 large-6 columns`,
  expressionPlotClassName: `small-12 large-6 columns`,
  speciesName: ``,
  height: 800,
  ks: []
}

export default TSnePlotWidget
