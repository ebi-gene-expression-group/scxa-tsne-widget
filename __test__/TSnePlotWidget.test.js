import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import TSnePlotWidget from '../src/TSnePlotWidget.js'
import TSnePlotView from 'expression-atlas-experiment-page-tsne-plot'

describe(`TSnePlotWidget`, () => {

  test(`weather showControl prop is true or false`, () => {
    const experimentAccession = `E-MTAB-5061`
    const speciesName = `Homo Sapiens`

    const onChangeColourBy = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const wrapper = mount(<TSnePlotWidget speciesName={speciesName} experimentAccession={experimentAccession} ks={[]} metadata={[]} selectedColourBy={`0`} onChangeColourBy={onChangeColourBy} perplexities={[]} selectedPerplexity={0} onChangePerplexity={onChangePerplexity} loading={true} plotData={plotData}/>)

    const tSnePlotView = wrapper.find(TSnePlotView)

    expect(tSnePlotView.props().showControls).toBe(false)
  })

  test(`matches snapshot`, () => {
    const experimentAccession = `E-MTAB-5061`
    const speciesName = `Homo Sapiens`

    const onChangeColourBy = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const tree = renderer
      .create(<TSnePlotWidget speciesName={speciesName} experimentAccession={experimentAccession} ks={[]} metadata={[]} selectedColourBy={`0`} onChangeColourBy={onChangeColourBy} perplexities={[]} selectedPerplexity={0} onChangePerplexity={onChangePerplexity} loading={true} plotData={plotData}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
