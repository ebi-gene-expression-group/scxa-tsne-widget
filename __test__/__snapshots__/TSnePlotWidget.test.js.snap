// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSnePlotWidget matches snapshot 1`] = `
Array [
  <div
    className="row expanded"
  >
    <div
      className="small-12 large-6 columns"
    >
      <div
        className="row"
      >
        <div
          className="small-12 medium-6 columns"
        >
          <label>
            Colour plot by:
          </label>
          <div
            className=" css-2b097c-container"
            onKeyDown={[Function]}
          >
            <div
              className=" css-tinoic"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className=" css-1hwfws3"
              >
                <div
                  className=" css-1wa3eu0-placeholder"
                >
                  Select...
                </div>
                <div
                  className="css-1g6gooi"
                >
                  <div
                    className=""
                    style={
                      Object {
                        "display": "inline-block",
                      }
                    }
                  >
                    <input
                      aria-autocomplete="list"
                      autoCapitalize="none"
                      autoComplete="off"
                      autoCorrect="off"
                      disabled={false}
                      id="react-select-4-input"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      spellCheck="false"
                      style={
                        Object {
                          "background": 0,
                          "border": 0,
                          "boxSizing": "content-box",
                          "color": "inherit",
                          "fontSize": "inherit",
                          "label": "input",
                          "opacity": 1,
                          "outline": 0,
                          "padding": 0,
                          "width": "1px",
                        }
                      }
                      tabIndex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style={
                        Object {
                          "height": 0,
                          "left": 0,
                          "overflow": "scroll",
                          "position": "absolute",
                          "top": 0,
                          "visibility": "hidden",
                          "whiteSpace": "pre",
                        }
                      }
                    >
                      
                    </div>
                  </div>
                </div>
              </div>
              <div
                className=" css-1wy0on6"
              >
                <span
                  className="sc-bdVaJa hUYIaS"
                  cx={[Function]}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="row text-center scxa-error"
      >
        <p>
          ReferenceError: fetch is not defined
        </p>
      </div>
    </div>
    <div
      className="small-12 large-6 columns"
    >
      <div
        style={
          Object {
            "visibility": "hidden",
          }
        }
      >
        <div
          className="row margin-bottom-large"
        >
          <label>
            Gene ID, gene name or gene feature
          </label>
          <div
            style={
              Object {
                "display": "",
              }
            }
          >
            <input
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              name="geneId"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              role="combobox"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="row text-center scxa-error"
      >
        <p>
          ReferenceError: fetch is not defined
        </p>
      </div>
    </div>
  </div>,
  <p>
    To know more about this experiment please go to 
    <a
      href="https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-5061"
      target="_blank"
    >
      Single Cell Expression Atlas 
    </a>
    .
  </p>,
]
`;
